Add AddSystemd() and AddWindowsService() IServiceCollection extension methods

halter73 opened the pull request:
This finishes the work for #61634 which tracked adding `HostApplicationBuilder`.

This should also resolve #69212 once completed.

halter73 committed:
Add UseSystemd() IServiceCollection extension method

halter73 commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/UseWindowsServiceTests.cs:
```diff
@@ -2,26 +2,159 @@
+                    //   at Microsoft.Extensions.Hosting.WindowsServices.WindowsServiceLifetime..ctor(IHostEnvironment environment, IHostApplicationLifetime applicationLifetime, ILoggerFactory loggerFactory, IOptions`1 optionsAccessor, IOptions`1 windowsServiceOptionsAccessor) in C:\dev\dotnet\runtime\src\libraries\Microsoft.Extensions.Hosting.WindowsServices\src\WindowsServiceLifetime.cs:line 26
+
+                    // REVIEW: This seems similar to https://github.com/dotnet/sdk/issues/16049 which was closed. I'm not sure why this is still happening in tests on Windows.
+                    // net462 tests can construct ServiceBase just fine, but not net7.0. Does anyone know what the real issue is here?
```
@eerhardt Do you have any idea what's going on here? I see you commented on https://github.com/dotnet/sdk/issues/16049.

eerhardt replied:
My initial guess at the issue here is that System.ServiceProcess.ServiceController targets `net7.0-windows` and `net7.0`.

https://github.com/dotnet/runtime/blob/7654bf2c63d22dfdf197d54c1f4b7017aa556726/src/libraries/System.ServiceProcess.ServiceController/src/System.ServiceProcess.ServiceController.csproj#L3

The `windows`-specific version is what has the real code in it. The platform-agnostic version has PNSE all over, because you can't do windows things on platform-agnostic targets.

The tests only target `net7.0`. So when we build the tests, it picks the platform-agnostic version of `System.ServiceProcess.ServiceController`, which PNSEs.

So the fix here would be to target the tests for `windows` specifically. Check out here as an example:

https://github.com/dotnet/runtime/blob/7654bf2c63d22dfdf197d54c1f4b7017aa556726/src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.Tests.csproj#L2-L4

That's if you just want the tests to run on windows. If you want to run on all platforms, use:

https://github.com/dotnet/runtime/blob/7654bf2c63d22dfdf197d54c1f4b7017aa556726/src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Microsoft.Extensions.Hosting.Unit.Tests.csproj#L3-L4

halter73 commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/UseWindowsServiceTests.cs:
```diff
@@ -2,26 +2,159 @@
+                ?? throw new MissingMethodException();
+
+            configure ??= _ => { };
+            _useWindowsServiceUncheckedMethod.Invoke(null, new object[] { services, configure });
```
This is the right pattern for testing non-public APIs rather than InternalsVisibleTo, right?

eerhardt replied:
Correct. We opt for reflection over IVT.

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -30,6 +30,17 @@ public WindowsServiceLifetime(IHostEnvironment environment, IHostApplicationLife
+            if (windowsServiceOptionsAccessor.Value.ValidateContentRoot && !string.Equals(environment.ContentRootPath, AppContext.BaseDirectory, StringComparison.Ordinal))
+            {
+                throw new InvalidOperationException(
+                    $"The IHostEnvironment.ConentRootPath value of '{environment.ContentRootPath}' must match the AppContext.BaseDirectory value of '{AppContext.BaseDirectory}' but is unequal.");
```
Do we not (yet) use resources for such error messages in Microsoft.Extensions libs?

eerhardt replied:
>  Some of the bigger Microsoft.Extensions libs have already had resources added.

Yeah, we've been adding them as we go. So for new strings, I would vote to put them in a .resx from the start.

martincostello commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetimeOptions.cs:
```diff
@@ -11,5 +11,11 @@ public class WindowsServiceLifetimeOptions
         /// </summary>
         public string ServiceName { get; set; } = string.Empty;
+
+        // Used by the IServiceCollection overload of UseWindowsService to indicated that WindowsServiceLifetime
```
Nit:

```suggestion
        // Used by the IServiceCollection overload of UseWindowsService to indicate that WindowsServiceLifetime
```

martincostello commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHostBuilderExtensions.cs:
```diff
@@ -27,27 +27,63 @@ public static class SystemdHostBuilderExtensions
+        ///   </para>
+        ///   <para>
+        ///     The systemd service file must be configured with <c>Type=notify</c> to enable
+        ///     notifications. See https://www.freedesktop.org/software/systemd/man/systemd.service.html.
```
Would be good to do the same as https://github.com/dotnet/aspnetcore/pull/41254

```suggestion
        ///     notifications. See <see href="https://www.freedesktop.org/software/systemd/man/systemd.service.html"/>.
```

halter73 commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -30,6 +30,17 @@ public WindowsServiceLifetime(IHostEnvironment environment, IHostApplicationLife
+            // The IHostBuilder overload sets the content root unlike the IServiceCollection overload, so it should be less likely to be wrong anyway.
+            // We could try normalizing the paths before comparing, but if the ContentRootPath has changed from it's AppContext.BaseDirectory default,
+            // there's probably a misconfiguration.
+            if (windowsServiceOptionsAccessor.Value.ValidateContentRoot && !string.Equals(environment.ContentRootPath, AppContext.BaseDirectory, StringComparison.Ordinal))
```
This is indeed a problem. The CWD is the default, and this is the C:\Windows\system32 folder for Windows services. `AppContext.BaseDirectory` is the `DisableDefaults` behavior and is what we want. C:\Windows\system32 is not a reasonable content root, so we have to do something.

We can change the exception message reference `HostApplicationBuilderSettings.DisableDefaults`, but we cannot be sure if this is extension method is being used with a `HostApplicationBuilder`. The other problem is that it's not always the easiest to observe exceptions thrown from Windows Services.

I think we might have to add a `CreateWindowsServiceApplicationBuilder(WindowsServiceSettings)` method that returns a `HostApplicationBuilder` preconfigured with the right content root, lifetime and loggers. I'm not sure if we would want to follow a similar pattern for systemd even though we don't need to change the content root just to be consistent.

@Tratcher Do you have any ideas? The fundamental problem is that we cannot change the content root of `HostApplicationBuilder` after it's constructed in `UseWindowsService()`.

More context:

- https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-6.0&tabs=visual-studio#current-directory-and-content-root
- https://github.com/dotnet/extensions/pull/1083#issuecomment-462994671

davidfowl replied:
Autodetect when running in a service then change the default. That's the most user friendly thing to do.

eerhardt replied:
> The fundamental problem is that we cannot change the content root of HostApplicationBuilder after it's constructed

Maybe we should examine if this limitation is strictly necessary. What would be the harm if this was mutable? Yes someone could have read it before it was modified, but that's how every normal .NET object works.

Tratcher replied:
> I think we might have to add a `CreateWindowsServiceApplicationBuilder(WindowsServiceSettings)` method that returns a `HostApplicationBuilder` preconfigured with the right content root, lifetime and loggers. I'm not sure if we would want to follow a similar pattern for systemd even though we don't need to change the content root just to be consistent.

That doesn't seem to combine well. UseWindowsService was supposed to light up when running in that environment, having to create a separate builder defeats that.

Can we put an extension method UseWindowsServiceSettings on HostApplicationBuilderSettings and have it conditionally set the ContentRootPath? Then you can recommend that in this error message.

halter73 replied:
> > The fundamental problem is that we cannot change the content root of HostApplicationBuilder after it's constructed
> 
> Maybe we should examine if this limitation is strictly necessary. What would be the harm if this was mutable? Yes someone could have read it before it was modified, but that's how every normal .NET object works.

By the time the `HostApplicationBuilder` is constructed, we've already read `appsettings.json` and bootstrapped our environment. We'd have to update `JsonConfigurationSource`/`FileConfigurationSource` to stop caching the `IFileProvider` and/or, make `PhysicalFileProvider` rerootable. Handling the `IChangeToken`'s might be tricky with a rerootable `PhysicalFileProvider`, but it's probably possible. We'd also have to make sure `HostApplicationBuilder.Environment` gets updated.

The other question is what would the API for updating the content root be? It can't be modifying `HostApplicationBuilder.Configuration[HostDefaults.ContentRootKey]` because that would create a circular dependency. We could consider reacting when apps invoke the `Environment.EnvironmentName/ApplicationName/ContentRootPath/ContentRootFileProvider` setters, but allowing the `IFileProvider` itself to be updated means we would have to stop caching at the `FileConfigurationSource` level not just the `PhysicalFileProvider` level.

So far with `WebApplicationBuilder`, we've prevented the modification of the `WebRootPath` in addition to the `ContentRootPath` and the `WebRootPath` is usually dependent on the `ContentRootPath`. We'd have to think about how to rebase the `WebRootPath` when the `ContentRootPath` updates, but only if it was kept default or set to a relative path.  We'd also need to rerun the default `UseStaticWebAssets()` logic. This way leads to madness. See https://github.com/dotnet/aspnetcore/pull/37033 and eventually https://github.com/dotnet/aspnetcore/pull/37147 for prior attempts a dealing with this.

Another smaller thing to consider is `IHostingStartup`s. Today, the run as part of the `WebApplicationBuilder` constructor. This is important because it can do things like add configuration sources visible to `WebApplicationBuilder.Configuration`. These  `IHostingStartup` are guaranteed  to read the environment before the app has a chance to modify it unless the developer really goes out of the way to disable running these automatically and run them manually.

I really would like to remove this limitation because it is really unfortunate for scenarios such as this, but I'm not sure it's worth all the complexity it would bring to the implementation and how hard it would make it to reason about the behavior.

> That doesn't seem to combine well. UseWindowsService was supposed to light up when running in that environment, having to create a separate builder defeats that.

Agreed.

> Can we put an extension method UseWindowsServiceSettings on HostApplicationBuilderSettings and have it conditionally set the ContentRootPath? Then you can recommend that in this error message.

We could. This still isn't great because it's not the easiest to observe exceptions thrown from Windows Services before the `EventLogLoggerProvider` is configured. It might also be confusing if called on an `IServiceCollection` belonging to an `HostBuilder` instead of an `HostApplicationBuilder`.

> Autodetect when running in a service then change the default. That's the most user friendly thing to do.

This is my preference given the option. This would mean type forwarding almost the entirety of `Microsoft.Extensions.Hosting.WindowsServices` and `System.ServiceProcess.ServiceController` to `Microsoft.Extensions.Hosting`. Are we okay with that?

davidfowl replied:
> Autodetect when running in a service then change the default. That's the most user friendly thing to do.

No I mean specifically for the content root, when running on windows, using systems APIs to detect (so no dependency) and changing the content root to be the app base instead.

Basically, if the CWD is the "System folder" then change it to be the app base.

davidfowl replied:
Windows services run in a shared process and CWD is `C:\WINDOWS\system32` or SystemDrive\Windows\System32. This is no different from what we do today to [turn on the event log for errors by default on windows](https://github.com/dotnet/runtime/blob/8823eefe91f6b774961c32800dea977b687d7328/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L253-L257)

halter73 replied:
Could we use `WindowsServiceHelpers.IsWindowsService()` and by extension `Win32.GetParentProcess()` to determine whether or not to use the CWD or `AppContext.BaseDirectory`? Or do you think it's really better to compare the CWD against the system32 directory? We could just share the source.

halter73 replied:
And would we make the same change to `HostBuilder`?

davidfowl replied:
> Could we use WindowsServiceHelpers.IsWindowsService() and by extension Win32.GetParentProcess() to determine whether or not to use the CWD or AppContext.BaseDirectory? Or do you think it's really better to compare the CWD against the system32 directory? We could just share the source.

I was thinking about the most robust way to do this. Maybe checking for windows services is cleaner but less general purpose. I think we should do this in the "default" host builder scenario.

halter73 replied:
> After normalizing the path of course (PS: I wish we had that built in with strongly typed Path objects).

I really wish we had a reliable way to compare path equality. I get this is a hard problem with various filesystems affecting how paths are normalized. I imagine hard and soft links don't make things any easier particularly if the path points to a file or directory that hasn't been created yet. That's why it would be nice to have an API instead of trying to do this ourselves like we did in `WebApplicationBuilder`.

davidfowl replied:
Agree. I think for now you'll have to do something similar to what's in `WebApplicationBuilder`

ericstj replied:
There's no fool proof path comparison, you can even have a filesystem or hardware hide this information, you can run into opaque links (imagine multiple machines and network shares, mapped drives, etc), or permissions which prevent tracing all the paths.  Best to approximate with normalization then comparison.  Some platforms/filesystems could permit a "better" comparison (usually by opening the file and asking the filesystem for more information about it) - but you'll never get it 100% and the cost of getting higher than an approximation is usually too high for the value you get and has side effects (sharing / access denied / etc).  cc @adamsitnik @JeremyKuhne

eerhardt replied:
You can open the issue here:

https://github.com/dotnet/docs/issues/new?template=breaking-change.yml

That will then get converted into documentation.

eerhardt replied:
I ended up writing it up here: https://github.com/dotnet/docs/issues/29945. Please review and edit as necessary.

ericstj commented:
@halter73 are you planning to drive this further?

halter73 commented:
Yes. I'm busy with some stuff in the aspnetcore repo at the moment, but I plan to get to this later in the week. Hopefully, I'll have an updated PR by Wednesday.

halter73 committed:
Add UseWindowsService() IServiceCollection extension method

halter73 committed:
Target Windows for UseWindowsServiceTests

halter73 committed:
Stop throwing InvalidOperationExceptions from WindowsServiceLifetime

halter73 committed:
Don't default to CWD if in C:\Windows\system32

halter73 committed:
Use* to Add* renames

halter73 committed:
Apply suggestions from code review

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>
Co-authored-by: Martin Costello <martin@martincostello.com>

halter73 commented on src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs:
```diff
@@ -197,10 +197,20 @@ public static IHostBuilder ConfigureDefaults(this IHostBuilder builder, string[]
+            // any trailing directory separator characters. I'm not even sure the casing can ever be different from these APIs, but I it makes sense to
+            // ignore case for Windows path comparisons given the file system is usually (always?) going to be case insensitive for the system path.
+            var cwd = Environment.CurrentDirectory;
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) || !string.Equals(cwd, Environment.GetFolderPath(Environment.SpecialFolder.System), StringComparison.OrdinalIgnoreCase))
```
This PR has now been updated so hosting never sets the CWD as the content root if it's C:\Windows\system32 even if you explicitly call `IHostBuilder.ConfigureDefaults` or you use a `HostApplicationBuilder` without `DisableDefaults`.

You can still manually set this as the `ContentRoot` if you want to, but this will need to be announced as a breaking change.

davidfowl replied:
What will the content root default to?

halter73 replied:
Unless explicitly set to something out, the truly-default default 😆 content root for hosting is  `AppContext.BaseDirectory` which is what `UseWindowsService` used to set it to for you before this change.

davidfowl replied:
Great!

halter73 commented:
I think I've addressed all the feedback so far.

halter73 committed:
UseSystemd to AddSystemd ref fixup

halter73 committed:
UseWindowsService to AddWindowsService ref fixup

halter73 committed:
Remove extraneous code and outdated comments

halter73 committed:
Fix more outdated comments

halter73 committed:
delete more

halter73 committed:
Complete comment in ApplyDefaultHostConfiguration

halter73 committed:
Keep the tests hygienic

halter73 committed:
less var

halter73 committed:
Try fixing CreateDefaultBuilder_DoesNotChangeContentRootIfCurrentDirectoryIsWin

- It was failing on Windows.Nano.1809.Amd64.Open containers only.
- Was it silently failing to change the CWD?

halter73 commented:
Btw, I'll be OOF for the next two weeks. I realize this doesn't give a ton of time for a follow up review, but I think it's in a good state. The last thing I did was try to fix a test failure on a Windows.Nano.1809.Amd64.Open container. I don't have a good way to test locally so we'll see how it goes. I'm guessing it's silently failing to change the CWD to the system directory in the test.

I can find someone else to shepherd it through the rest of the way if need be.

eerhardt commented:
> I can find someone else to shepherd it through the rest of the way if need be.

I can shepherd this through while you are out. I'm reviewing it now.

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs:
```diff
@@ -197,10 +197,22 @@ public static IHostBuilder ConfigureDefaults(this IHostBuilder builder, string[]
+            // any trailing directory separator characters. I'm not even sure the casing can ever be different from these APIs, but I think it makes sense to
+            // ignore case for Windows path comparisons given the file system is usually (always?) going to be case insensitive for the system path.
+            string cwd = Environment.CurrentDirectory;
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) || !string.Equals(cwd, Environment.GetFolderPath(Environment.SpecialFolder.System), StringComparison.OrdinalIgnoreCase))
```
If we wanted to sanitize the path, we could follow this same approach:

[https://github.com/dotnet/runtime/blob/051b4828c7d3a0cad3289830ef9fd2120f45bb2b/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs#L299-L305](https://github.com/dotnet/runtime/blob/051b4828c7d3a0cad3289830ef9fd2120f45bb2b/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs#L299-L305)

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs:
```diff
@@ -46,6 +46,47 @@ public void CreateDefaultBuilder_IncludesContentRootByDefault()
+
+            try
+            {
+                Environment.CurrentDirectory = systemDirectory;
```
This should probably use RemoteExecutor so it doesn't affect other tests that may be relying on CurrentDirectory.

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Microsoft.Extensions.Hosting.Unit.Tests.csproj:
```diff
@@ -7,6 +7,13 @@
   </PropertyGroup>
 
+  <ItemGroup>
+    <!-- Disable parallelization because CreateDefaultBuilder_DoesNotChangeContentRootIfCurrentDirectoryIsWindowsSystemDirectory changes the CWD. -->
```
Use RemoteExecutor instead. If needed, I can provide an example to copy.

halter73 replied:
An example to copy would be appreciated!

eerhardt commented:
This looks pretty good. I just had a few comments. If you don't have time to resolve them before you are OOF, I can do it early next week. Just let me know.

Thanks again for pushing on this.

halter73 commented:
> I can shepherd this through while you are out. I'm reviewing it now. I'd like to get it in for the next preview release.

Thanks @eerhardt! I think, I'll take you up on your offer.

halter73 committed:
What's really going on?

halter73 committed:
Use RemoteExecutor

halter73 commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs:
```diff
@@ -46,6 +48,45 @@ public void CreateDefaultBuilder_IncludesContentRootByDefault()
+                    // Maybe this doesn't happen anymore if the container doesn't support RemoteExecutor.
+                    if (config["ContentRoot"] is not null)
+                    {
+                        throw new Exception($"ContentRoot: {config["ContentRoot"]}, originalCwd: {originalCwd}, systemDirectory: {systemDirectory}, Environment.CurrentDirectory: {Environment.CurrentDirectory}, Environment.SpecialFolder.System: {Environment.SpecialFolder.System}");
```
```suggestion
                        throw new Exception($"ContentRoot: {config["ContentRoot"]}, originalCwd: {originalCwd}, systemDirectory: {systemDirectory}, Environment.CurrentDirectory: {Environment.CurrentDirectory}, Environment.SpecialFolder.System: {Environment.GetFolderPath(Environment.SpecialFolder.System)}");
```

halter73 replied:
It looks like the Windows container doesn't have a system directory. I guess we could just skip this test if `string.IsNullOrEmpty(Environment.GetFolderPath(Environment.SpecialFolder.System))`. We can just add to the already-verbose skip condition: `IsWindowsAndHasSystemFolderAndRemotExecutorIsSupported`😆

eerhardt replied:
Looks like this is caused by https://github.com/dotnet/runtime/issues/21430. I disabled the test when `Environment.GetFolderPath(Environment.SpecialFolder.System)` returns null or empty.

halter73 committed:
Update src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHostBuilderExtensions.cs:
```diff
@@ -27,27 +27,63 @@ public static class SystemdHostBuilderExtensions
+        /// </param>
+        /// <returns>The <paramref name="services"/> instance for chaining.</returns>
+        public static IServiceCollection AddSystemd(this IServiceCollection services)
+        {
```
In extensions, so we typically null check receivers to throw an ArgumentNullException instead of null ref'ing? If no, this is fine. If yes, we should here, too.

eerhardt replied:
It's pretty inconsistent. Sometimes we don't check:

https://github.com/dotnet/runtime/blob/807d0916c5205d75ecb04809f8106a3caa46bbe4/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L30-L86

https://github.com/dotnet/runtime/blob/807d0916c5205d75ecb04809f8106a3caa46bbe4/src/libraries/Microsoft.Extensions.Logging/src/LoggingBuilderExtensions.cs#L22-L62

But for all `IServiceCollection` args we do seem to check:

https://github.com/dotnet/runtime/blob/807d0916c5205d75ecb04809f8106a3caa46bbe4/src/libraries/Microsoft.Extensions.Logging/src/LoggingServiceCollectionExtensions.cs#L32-L34

https://github.com/dotnet/runtime/blob/807d0916c5205d75ecb04809f8106a3caa46bbe4/src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/src/OptionsConfigurationServiceCollectionExtensions.cs#L62-L65

So I'll add them here.

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/tests/UseSystemdTests.cs:
```diff
@@ -12,16 +12,28 @@ public class UseSystemdTests
+            using IHost host = builder.Build();
+
+            var lifetime = host.Services.GetRequiredService<IHostLifetime>();
+            Assert.IsNotType<SystemdLifetime>(lifetime);
```
What type will it be? If it's actually going to be null, we should validate that instead.

halter73 replied:
It's usually `ConsoleLifetime`. `UseWindowsServiceTests.DefaultsToOffOutsideOfService` does an `Assert.IsType<ConsoleLifetime>(lifetime)` check like you recommend, so I'm not sure why it was different here. Maybe because the `UseSystemdTests` also run on some of the platforms with the `NullLifetime`?

https://github.com/dotnet/runtime/blob/9d0fcb5c2030f2a394eb4d0e675ba0de6cac72d1/src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.netcoreapp.cs#L14-L21

eerhardt replied:
That's correct - these tests run on Android, iOS, and WASM. So it it either `ConsoleLifetime` or `NullLifetime`. This test is just making sure it isn't `SystemdLifetime`, since it isn't in a systemd process.

I've added back the `Assert.NotNull`, since as far as I can tell if it is `null`, the test would pass:

https://github.com/xunit/assert.xunit/blob/58a59628853ed7a63895945fdac7984a92da2343/TypeAsserts.cs#L61-L85

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHostBuilderExtensions.cs:
```diff
@@ -27,27 +27,63 @@ public static class SystemdHostBuilderExtensions
+            });
+
+            // IsSystemdService() will never return true for android/browser/iOS/tvOS
+#pragma warning disable CA1416 // Validate platform compatibility
```
The Windows service code below uses asserts for suppression. Should we do the same here instead of pragmas and comments?

eerhardt replied:
The issue here is that the unix condition is harder to write. The assert for windows is that it "is windows". The assert here would need to be not any of the following:

*    [UnsupportedOSPlatform("android")]
*    [UnsupportedOSPlatform("browser")]
*    [UnsupportedOSPlatform("ios")]
*    [UnsupportedOSPlatform("maccatalyst")]
*    [UnsupportedOSPlatform("tvos")]

Which isn't that bad. But where it does get bad is that you can't express this condition outside of NETCOREAPP since we don't have APIs in `RuntimeInformation` and the `OperatingSystem.IsXXX` methods are only available on NETCOREAPP.

So we need to pragma disable here.

eerhardt committed:
Skip test on Windows nano server

eerhardt committed:
Respond to PR feedback

eerhardt commented:
Breaking change doc created here: https://github.com/dotnet/docs/issues/29945
